lib/pat.c

```diff
@@ -5990,18 +5990,64 @@ pat_key_defrag_each(grn_ctx *ctx,
   return new_curr_key;
 }
 
+static inline grn_rc
+scan_delinfos_and_phase2(grn_ctx *ctx, grn_pat *pat)
+{
+  for (uint32_t di_index = 0; di_index < GRN_PAT_NDELINFOS; di_index++) {
+    grn_pat_delinfo *delete_info = &(pat->header->delinfos[di_index]);
+    if (delete_info->stat != DL_PHASE1) {
+      continue;
+    }
+
+    grn_pat_wal_add_entry_data wal_data = {0};
+    wal_data.tag = "[pat][defrag][delinfos][phase2]";
+    wal_data.pat = pat;
+    wal_data.delete_info_index = di_index;
+    grn_rc rc = delinfo_turn_2(ctx, pat, &wal_data, di_index);
+    if (rc != GRN_SUCCESS) {
+      ERR(rc, "%s: failed %d", wal_data.tag, delete_info->ld);
+      return rc;
+    }
+    delinfo_turn_2_post(ctx, pat);
+  }
+  return GRN_SUCCESS;
+}
+
+/* Clear to avoid reuse of unexpected nodes. */
+static inline void
+pat_clear_garbages(grn_pat *pat)
```

Comment:
あとあと、`ctx`を使ってなにかしたい時に追加するとインターフェースが変わってしまうので、最初から`ctx`を渡すようにしておく方が良いかなと思いました。（`static`な関数なのでインターフェースを変更しても影響は限定的だとは思うのですが。）

```suggestion
pat_clear_garbages(grn_ctx ctx, grn_pat *pat)
```